### ISSUE TITLE
Which kind of input is better for LSTM?

### ISSUE THREAD

**Issue by chooron** [neuralhydrology/neuralhydrology#92]
Hi, I recently encountered some problems when using LSTM to realize rainfall runoff prediction. In general, the input to the LSTM model is [**batch_size, silde_window, feature_dim**], but I also find that [**batch_size, 1, silde_window * feature_dim**] can also make robust forecast, and the model training time is shorter. There is my lstm code.
```
class RNN_Base(nn.Module):
    _cell = nn.LSTM
    def __init__(self, c_in, c_out, hidden_size=100, n_layers=1, bias=True, rnn_dropout=0, bidirectional=False, fc_dropout=0., is_reshape=False, seq_len=None):
        super().__init__()
        self.is_reshape = is_reshape
        if self.is_reshape:
            self.seq_len = seq_len
            self.rnn = self._cell(c_in * seq_len, hidden_size, num_layers=n_layers, bias=bias, batch_first=True, dropout=rnn_dropout, bidirectional=bidirectional)
            self.flatten = Reshape(-1)
        else:
            self.rnn = self._cell(c_in, hidden_size, num_layers=n_layers, bias=bias, batch_first=True, dropout=rnn_dropout, bidirectional=bidirectional)
        self.dropout = nn.Dropout(fc_dropout)
        self.fc = nn.Linear(hidden_size * (1 + bidirectional), c_out)
        self.apply(self._weights_init)

    def forward(self, x):
        # x [batch_size , silde_window , feature_dim]
        if self.is_reshape: # this control the different input type
            x = self.flatten(x) # convert to [batch_size , silde_window x feature_dim]
            x = x.unsqueeze(1) # convert to [batch_size ,1 , silde_window x feature_dim]
        output, _ = self.rnn(x)
        output = output[:, -1]
        output = self.fc(self.dropout(output))
        return output
```
From my perspective, [batch_size, 1, silde_window * feature_dim] convert the model input from matrix to vector, which may loss the time information in the matrix. However, when I compare the two different input type, [batch_size, 1, silde_window * feature_dim] did not show the problem of missing information. 
Thus, I sincerely ask you about the difference between these two LSTM inputs, and I want to know which input form is better for LSTM.
(My English ability is limited, I hope what I said did not offend you)

**Comment by kratzert** [comment]
Hi,

thanks for the question, even though this is not really the correct place (Issues are meant for code bugs/feature requests, etc.) and the discussion forum or google groups would be a better place for asking such questions.

If I understand correctly and without diving into your code you are asking me on my opinion on whether it is better to use the LSTM in a sequential mode (time step by time step) vs passing all temporal data flattened at once to the LSTM (i.e. the LSTM only performs a single time step). In that regard, my answer is pretty clear that if you want to use the LSTM, you should use it time step by timestep, otherwise there is no purpose on using the LSTM. The idea of the LSTM is that you input a sequence of data and the model can learn to store information over time and how the inputs/processes evolve over time. If you pass everything at one timestep, there are no temporal processes and the cell states have basically no effect. 
This has nothing to do with the performance by itself but I think if you want to pass all information at one timestep, why not use a fully connected network directly?
